core/data/tools_cli.joke

```diff
@@ -0,0 +1,706 @@
+
+(defn- end-of-args? [x]
+  (= "--" x))
+
+(defn- spec-for
+  [arg specs]
+  (->> specs
+       (filter (fn [s]
+                   (let [switches (set (s :switches))]
+                     (contains? switches arg))))
+       first))
+
+(defn- default-values-for
+  [specs]
+  (reduce (fn [m s]
+            (if (contains? s :default)
+              ((:assoc-fn s) m (:name s) (:default s))
+              m))
+          {} specs))
+
+(defn- apply-specs
+  [specs args]
+  (loop [options    (default-values-for specs)
+         extra-args []
+         args       args]
+    (if-not (seq args)
+      [options extra-args]
+      (let [opt  (first args)
+            spec (spec-for opt specs)]
+        (cond
+         (end-of-args? opt)
+         (recur options (into extra-args (vec (rest args))) nil)
+
+         (and (opt? opt) (nil? spec))
+         (throw (ex-info (str "'" opt "' is not a valid argument") {:opt opt}))
+
+         (and (opt? opt) (spec :flag))
+         (recur ((spec :assoc-fn) options (spec :name) (flag-for opt))
+                extra-args
+                (rest args))
+
+         (opt? opt)
+         (recur ((spec :assoc-fn) options (spec :name) ((spec :parse-fn) (second args)))
+                extra-args
+                (drop 2 args))
+
+         :default
+         (recur options (conj extra-args (first args)) (rest args)))))))
+
+(defn- switches-for
+  [switches flag]
+  (-> (for [^String s switches]
+        (cond (and flag (flag? s))
+              [(s/replace s #"\[no-\]" "no-") (s/replace s #"\[no-\]" "")]
+
+              (and flag (s/starts-with? s "--"))
+              [(s/replace s #"--" "--no-") s]
+
+              :default
+              [s]))
+      flatten))
+
+(defn- generate-spec
+  [raw-spec]
+  (let [[switches raw-spec] (split-with #(and (string? %) (opt? %)) raw-spec)
+        [docs raw-spec]     (split-with string? raw-spec)
+        options             (apply hash-map raw-spec)
+        aliases             (map name-for switches)
+        flag                (or (flag? (last switches)) (options :flag))]
+    (merge {:switches (switches-for switches flag)
+            :docs     (first docs)
+            :aliases  (set aliases)
+            :name     (keyword (last aliases))
+            :parse-fn identity
+            :assoc-fn assoc
+            :flag     flag}
+           (when flag {:default false})
+           options)))
+
+(defn- normalize-args
+  "Rewrite arguments sequence into a normalized form that is parsable by cli."
+  [specs args]
+  (let [required-opts (->> specs
+                           (filter (complement :flag))
+                           (mapcat :switches)
+                           (into #{}))
+        ;; Preserve double-dash since this is a pre-processing step
+        largs (take-while (partial not= "--") args)
+        rargs (drop (count largs) args)
+        [opts largs] (tokenize-args required-opts largs)]
+    (concat (mapcat rest opts) largs rargs)))
+
+(defn cli
```

Comment:
I think we should not bring over legacy API.

Reply:
Sounds good to me 👍